trusted-pgx/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+pub mod prelude {
+    pub use super::*;
+}
+
+pub use ::pgx::{debug1, debug2, debug3, debug4, debug5, ereport, error, info, notice, warning};
+
+pub use datum::*;
+pub mod datum {
+    pub use ::pgx::datum::{
+        AnyNumeric, Date, FromDatum, IntoDatum, Json, JsonB, Time, TimeWithTimeZone, Timestamp,
+        TimestampWithTimeZone,
+    };
+}
+
+pub mod fcinfo {
+    pub use ::pgx::fcinfo::pg_getarg;
+    pub use ::pgx::fcinfo::pg_return_null;
+    pub use ::pgx::fcinfo::pg_return_void;
+    pub use ::pgx::fcinfo::srf_first_call_init;
+    pub use ::pgx::fcinfo::srf_is_first_call;
+    pub use ::pgx::fcinfo::srf_per_call_setup;
+    pub use ::pgx::fcinfo::srf_return_done;
+    pub use ::pgx::fcinfo::srf_return_next;
+}
+
+pub use heap_tuple::*;
+pub mod heap_tuple {
+    pub use ::pgx::heap_tuple::PgHeapTuple;
+}
+
+pub use iter::*;
+pub mod iter {
+    pub use ::pgx::iter::{SetOfIterator, TableIterator};
+}
+
+pub use memcxt::*;
+pub mod memcxt {
+    pub use ::pgx::memcxt::PgMemoryContexts;
```

Comment:
I'm trying to reason if there's anything concerning with allowing regular users to manipulate memory contexts directly. Worst case they mess up their own allocations and drop potential allocations they might rely on later..? 

I guess it can be used to palloc large amounts of memory directly leading to stability issues?

Reply:
It’s required for pgx in general. 

We can audit PgMemoryContexts and make sure the things that should be “unsafe” are. And if that means every function that’s probably okay. 

Or I can implement a pared down version just for trusted-pgx.

Reply:
The latter already can be done via PL/Perl: simply make a bunch of arrays and fill them with meaningless junk data. I expect the same is true of PL/pgSQL though it might require something slightly more clever due to it being more heavily query-oriented.

Another good "joke" in that vein, this time in Rust, is to try `vec![1; 4_294_967_295]`.

Mind, any functions that allow them to actually violate memory safety *otherwise* should be marked `unsafe`.

Reply:
> Mind, any functions that allow them to actually violate memory safety otherwise should be marked unsafe.

yup. That’s really plrust’s only concern

Reply:
I'm working on this over on https://github.com/tcdi/pgx/pull/995.  That should get merged into pgx's develop branch later today, and then plrust's "unsafe" guards will simply prevent users from using PgMemoryContexts, and the other things in that PR.